components/tidb_query_datatype/src/codec/mysql/json/json_contains.rs

```diff
@@ -0,0 +1,94 @@
+// Copyright 2019 TiKV Project Authors. Licensed under Apache-2.0.
+
+use std::cmp::Ordering;
+
+use super::{super::Result, JsonRef, JsonType};
+
+impl<'a> JsonRef<'a> {
+    /// `json_contains` is the implementation for JSON_CONTAINS in mysql
+    /// <https://dev.mysql.com/doc/refman/5.7/en/json-search-functions.html#function_json-contains>
+    /// See `ContainsBinary()` in TiDB `json/binary_function.go`
+    pub fn json_contains(&self, target: JsonRef<'_>) -> Result<Option<bool>> {
+        match self.type_code {
+            JsonType::Object => {
+                if target.type_code == JsonType::Object {
+                    let elem_count = target.get_elem_count();
+                    for i in 0..elem_count {
+                        let key = target.object_get_key(i);
+                        let val = target.object_get_val(i)?;
+                        let idx = self.object_search_key(key);
+                        match idx {
+                            None => {
+                                return Ok(None);
+                            }
+                            Some(idx) => {
+                                let exp = self.object_get_val(idx)?;
+                                if exp.json_contains(val)? == None {
+                                    return Ok(None);
+                                }
+                            }
+                        }
+                    }
+                    return Ok(Some(true));
+                }
+            }
+            JsonType::Array => {
+                if target.type_code == JsonType::Array {
+                    let elem_count = target.get_elem_count();
+                    for i in 0..elem_count {
+                        if self.json_contains(target.array_get_elem(i)?)? == None {
+                            return Ok(None);
+                        }
+                    }
+                    return Ok(Some(true));
+                }
+                let elem_count = self.get_elem_count();
+                for i in 0..elem_count {
+                    if (self.array_get_elem(i)?.json_contains(target)?).is_some() {
+                        return Ok(Some(true));
+                    }
+                }
+            }
+            _ => {
+                return match self.partial_cmp(&target).unwrap() {
+                    Ordering::Equal => Ok(Some(true)),
+                    _ => Ok(None),
+                };
+            }
+        };
+        Ok(None)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::super::Json;
+    #[test]
+    fn test_json_contains() {
+        let mut test_cases = vec![
+            (r#"{"a":{"a":1},"b":2}"#, r#"{"b":2}"#, Some(true)),
```

Comment:
Add more tests:

```
            (r#"{}"#, r#"{}"#, Some(true)),
            (r#"{"a":1}"#, r#"{}"#, Some(true)),
            (r#"{"a":1}"#, r#"1"#, None),
            (r#"{"a":[1]}"#, r#"[1]"#, None),
            (r#"{"b":2, "c":3}"#, r#"{"c":3}"#, Some(true)),
            (r#"1"#, r#"1"#, Some(true)),
            (r#"[1]"#, r#"1"#, Some(true)),
            (r#"[1,2]"#, r#"[1]"#, Some(true)),
            (r#"[1,2]"#, r#"[1,3]"#, None),
            (r#"[1,2]"#, r#"["1"]"#, None),
            (r#"[1,2,[1,3]]"#, r#"[1,3]"#, Some(true)),
            (r#"[1,2,[1,[5,[3]]]]"#, r#"[1,3]"#, Some(true)),
            (r#"[1,2,[1,[5,{"a":[2,3]}]]]"#, r#"[1,{"a":[3]}]"#, Some(true)),
            (r#"[{"a":1}]"#, r#"{"a":1}"#, Some(true)),
            (r#"[{"a":1,"b":2}]"#, r#"{"a":1}"#, Some(true)),
            (r#"[{"a":{"a":1},"b":2}]"#, r#"{"a":1}"#, None),
```

As I have tried, they all passed.

Reply:
Added all these test and they all passed